src/Node.py

```diff
@@ -305,20 +308,40 @@ def get_messages(self):
         return messages
 
     def update_balloting_state(self, ballot):
-        """
-        Update the node's internal state based on quorum threshold checks for any of the prescribed state
-        """
-        if self.check_quorum_threshold_for_field(ballot, 'voted'):
-            self.balloting_state['voted'][ballot.value.hash] = ballot
-            log.node.info('Quorum threshold met for \'voted\' for ballot %s. Updated state', ballot)
-
-        if self.check_quorum_threshold_for_field(ballot, 'accepted'):
-            self.balloting_state['accepted'][ballot.value.hash] = ballot
-            log.node.info('Quorum threshold met for \'accepted\' for ballot %s. Updated state.', ballot)
-
-        if self.check_quorum_threshold_for_field(ballot, 'aborted'):
-            self.balloting_state['aborted'][ballot.value.hash] = ballot
-            log.node.info('Quorum threshold met for \'aborted\' for ballot %s. Updated state.', ballot)
+        ballot_hash = ballot.value.hash
+        if 'voted' not in self.balloting_state:
+            self.balloting_state['voted'] = {}
+        if 'accepted' not in self.balloting_state:
+            self.balloting_state['accepted'] = {}
+        if 'aborted' not in self.balloting_state:
+            self.balloting_state['aborted'] = {}
+        if ballot_hash not in self.balloting_state['voted']:
+            self.balloting_state['voted'][ballot_hash] = ballot
+            log.node.info('Node %s added ballot %s to voted state', self.name, ballot)
+        elif ballot_hash in self.balloting_state['voted']:
+            existing_ballot = self.balloting_state['voted'][ballot_hash]
+            if existing_ballot.counter < ballot.counter:
+                self.balloting_state['voted'][ballot_hash] = ballot
+                log.node.info('Node %s updated ballot %s in voted state with higher counter', self.name, ballot)
+
+        if ballot_hash in self.balloting_state['aborted']:
+            aborted_ballot = self.balloting_state['aborted'][ballot_hash]
+            if aborted_ballot.counter < ballot.counter:
+                self.balloting_state['aborted'][ballot_hash] = ballot
+                log.node.info('Node %s updated ballot %s in aborted state with higher counter', self.name, ballot)
```

Comment:
From this line (331) to 311, I don't think we need. As we are already instantiating the balloting_state in _innit__, we dont need to reinstantiate. Also, in the 'process_prepare_ballot_msg' which will be merged, we will already swap ballots with higher counters and deal with these cases. So this is not needed.

This function should just update a ballot from one state to another, to identify from which to which, it can be passes as a parameter. e.g. if 'voted' is passes as field, it will remove the ballot from 'voted' and update it to 'accepted'